### PR TITLE
Move Travis to container based infrastructure

### PR DESCRIPTION
**Pull request by thedrow**
This will allow us to cache the compilation of the code that hasn't changed.
It should greatly speed up the build. Moreover, container based build jobs are faster to boot.

### CODE REVIEW

**Comment by thedrow** [comment]
Hmm travis is not starting with this one. I need to check why.

**Comment by thedrow** [comment]
The extra packages are still not installed.

**Comment by thedrow** [comment]
Seems like nothing is cached. Hold on.

**Comment by thedrow** [comment]
I need to configure ccache with clang. Since the build jobs are running on 12.04 the ccache version is a bit old.
See http://petereisentraut.blogspot.co.il/2011/05/ccache-and-clang.html
http://petereisentraut.blogspot.co.il/2011/09/ccache-and-clang-part-2.html
http://peter.eisentraut.org/blog/2014/12/01/ccache-and-clang-part-3/

**Comment by thedrow** [comment]
If this PR is merged it should supersede #503.

**Comment by thedrow** [comment]
@miloyip This is more or less ready for review.

**Comment by Kosta-Github** [comment]
What about moving the repeated cmake file code snippet into its own file and including it from the 3 cmake files?

**Comment by thedrow** [comment]
@Kosta-Github I'm not very experienced with CMake. Can you please explain in more detail?
Do you mean moving the ccache related stuff to /CMakeModules/ccache.cmake?

**Comment by Kosta-Github** [comment]
exactly

**Comment by thedrow** [comment]
I'm not really sure how to do that.

**Comment by thedrow** [comment]
@Kosta-Github Care to show me?

File: .travis.yml [hunk 1]
```diff
@@ -1,51 +1,114 @@
...
+        apt:
+          packages:
+            - doxygen
 
 before_script:
-#   hack to avoid Valgrind bug (https://bugs.kde.org/show_bug.cgi?id=326469),
-#   exposed by merging PR#163 (using -march=native)
+    #   hack to avoid Valgrind bug (https://bugs.kde.org/show_bug.cgi?id=326469),
+    #   exposed by merging PR#163 (using -march=native)
+    - ccache -s
```
Comment (pah) [inline .travis.yml]: Please keep the comment next to the hack itself (i.e. move the `ccache` line before the comment).

Comment (thedrow) [inline .travis.yml]: Done :)

**Comment by thedrow** [comment]
Anything else before we merge this?

**Comment by thedrow** [comment]
@miloyip How can we address @Kosta-Github's remark? I don't know how to do this myself.

**Comment by miloyip** [comment]
@thedrow Oh. I am curious that I should have merged this PR to master but actually did not. Why?
https://github.com/miloyip/rapidjson/commits/master/.travis.yml

**Comment by thedrow** [comment]
Umm wait what just happened? We should contact Github support.
I can create a new pull request if you want me to.

**Comment by miloyip** [comment]
Yes, please create a new pull request. I really would like to merge your hard work.
Besides, please note that I have recently committed 403115019b26b92e146dd0174438e8c79c688cdd for compiling C++11 support.

**Pull request merged**